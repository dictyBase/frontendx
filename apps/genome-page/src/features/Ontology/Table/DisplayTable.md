```jsx
const data = require("../mockData").data

<DisplayTable data={data} />
```
