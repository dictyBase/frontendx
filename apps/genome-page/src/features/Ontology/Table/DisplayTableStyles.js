const styles = (theme) => ({
  root: {
    width: "100%",
    overflowX: "auto",
  },
  row: {
    "&:nth-of-type(even)": {
      backgroundColor: theme.palette.background.default,
    },
  },
  link: {
    textDecoration: "none",
    color: "#4C5E81",
    "&:visited": {
      color: "#4C5E81",
    },
    "&:hover": {
      textDecoration: "underline",
    },
  },
  head: {
    backgroundColor: "#e6f2ff",
  },
  headerCell: {
    color: "#333",
    fontWeight: "600",
  },
})

export default styles
