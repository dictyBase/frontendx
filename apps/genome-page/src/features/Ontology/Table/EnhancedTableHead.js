// @flow
import React, { Component } from "react"
import { withStyles } from "@material-ui/core/styles"
import TableHead from "@material-ui/core/TableHead"
import TableRow from "@material-ui/core/TableRow"
import TableCell from "@material-ui/core/TableCell"
import TableSortLabel from "@material-ui/core/TableSortLabel"

const columnData = [
  {
    id: "goterm",
    label: "GO Term + Extension",
  },
  { id: "evidence_code", label: "Evidence" },
  { id: "with", label: "With" },
  {
    id: "publication",
    label: "Reference",
  },
  { id: "date", label: "Date" },
  { id: "assigned_by", label: "Source" },
]

const styles = () => ({
  head: {
    backgroundColor: "#DFE8F6",
  },
  headerCell: {
    color: "#333",
    fontWeight: "600",
  },
})

type Props = {
  /** Material-UI styling */
  classes: Object,
  /** The order to sort the column */
  order: string,
  /** The item to be ordered by */
  orderBy: string,
  /** Function for handling sorting */
  onRequestSort: Function,
}

/**
 * Enhanced table head component that allows for column sorting.
 */

class EnhancedTableHead extends Component<Props> {
  createSortHandler = property => event => {
    this.props.onRequestSort(event, property)
  }

  render() {
    const { order, orderBy, classes } = this.props

    return (
      <TableHead className={classes.head}>
        <TableRow>
          {columnData.map((column: Object) => {
            return (
              <TableCell
                key={column.id}
                className={classes.headerCell}
                sortDirection={orderBy === column.id ? order : false}>
                {/* $FlowFixMe */}
                <TableSortLabel
                  active={orderBy === column.id}
                  direction={order}
                  onClick={this.createSortHandler(column.id)}>
                  {column.label}
                </TableSortLabel>
              </TableCell>
            )
          }, this)}
        </TableRow>
      </TableHead>
    )
  }
}

export default withStyles(styles)(EnhancedTableHead)
