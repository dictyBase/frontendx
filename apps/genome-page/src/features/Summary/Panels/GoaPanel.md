```js
const data = require("./mockData").data
;<GoaPanel data={data} />
```
