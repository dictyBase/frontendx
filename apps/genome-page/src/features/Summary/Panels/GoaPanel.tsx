import React from "react"
import GoaPanelContent from "./GoaPanelContent"
import ItemDisplay from "common/components/panels/ItemDisplay"
import LeftDisplay from "common/components/panels/LeftDisplay"
import RightDisplay from "common/components/panels/RightDisplay"
import { GoAnnotation } from "dicty-graphql-schema"

type GOType = "molecular_function" | "biological_process" | "cellular_component"

const dataFilter = (arr: Array<GoAnnotation>, type: GOType) => {
  // get the attributes from specified type
  const attr = arr.filter((item) => item.type === type)

  // get the five most recent EXP annotations
  const expChecker = attr
    .filter(
      (item) =>
        item.evidence_code === "IMP" ||
        item.evidence_code === "IGI" ||
        item.evidence_code === "IDA" ||
        item.evidence_code === "IPI" ||
        item.evidence_code === "IEP" ||
        item.evidence_code === "EXP",
    )
    .sort((a, b) => b.date.localeCompare(a.date))
    .slice(0, 5)

  // get five most recent manual
  const manualChecker = attr
    .filter((item) => item.evidence_code !== "IEA")
    .sort((a, b) => b.date.localeCompare(a.date))
    .slice(0, 5)

  // get five most recent electronic
  const electronicChecker = attr
    .filter((item) => item.evidence_code === "IEA")
    .sort((a, b) => b.date.localeCompare(a.date))
    .slice(0, 5)

  // check if EXP array is empty
  // if it is, return manual
  // if manual is empty, return electronic
  if (!Array.isArray(expChecker) || !expChecker.length) {
    if (!Array.isArray(manualChecker) || !manualChecker.length) {
      return electronicChecker
    }
    return manualChecker
  }
  return expChecker
}

type Props = {
  /** Array of GO annotations for a particular gene */
  data: Array<GoAnnotation>
}

/**
 * Panel to display Gene Ontology Annotations on the Gene Summary page.
 */

const GoaPanel = ({ data }: Props) => (
  <div>
    <ItemDisplay>
      <LeftDisplay>Molecular Function</LeftDisplay>
      <RightDisplay>
        {dataFilter(data, "molecular_function").map(
          (item: GoAnnotation, index: number) => (
            <GoaPanelContent key={index} data={item} />
          ),
        )}
      </RightDisplay>
    </ItemDisplay>
    <ItemDisplay>
      <LeftDisplay>Biological Process</LeftDisplay>
      <RightDisplay>
        {dataFilter(data, "biological_process").map(
          (item: GoAnnotation, index: number) => (
            <GoaPanelContent key={index} data={item} />
          ),
        )}
      </RightDisplay>
    </ItemDisplay>
    <ItemDisplay>
      <LeftDisplay>Cellular Component</LeftDisplay>
      <RightDisplay>
        {dataFilter(data, "cellular_component").map(
          (item: GoAnnotation, index: number) => (
            <GoaPanelContent key={index} data={item} />
          ),
        )}
      </RightDisplay>
    </ItemDisplay>
  </div>
)

export default GoaPanel
