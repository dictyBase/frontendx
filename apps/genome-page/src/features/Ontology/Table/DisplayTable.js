// @flow
import React, { Component, Fragment } from "react"
import { withStyles } from "@material-ui/core/styles"
import Table from "@material-ui/core/Table"
import TableBody from "@material-ui/core/TableBody"
import TableCell from "@material-ui/core/TableCell"
import TableRow from "@material-ui/core/TableRow"
import Paper from "@material-ui/core/Paper"

import EnhancedTableHead from "./EnhancedTableHead"
import pubLinkGenerator from "../utils/pubLinkGenerator"
import evidenceLinkGenerator from "../utils/evidenceLinkGenerator"
import withLinkGenerator from "../utils/withLinkGenerator"
import removeUnderscores from "../utils/removeUnderscores"
import dateConverter from "../utils/dateConverter"
import qualifierFormatter from "../utils/qualifierFormatter"

const styles = (theme: Object) => ({
  root: {
    width: "100%",
    overflowX: "auto",
  },
  row: {
    "&:nth-of-type(even)": {
      backgroundColor: theme.palette.background.default,
    },
  },
  link: {
    textDecoration: "none",
    color: "#4C5E81",
    "&:visited": {
      color: "#4C5E81",
    },
    "&:hover": {
      textDecoration: "underline",
    },
  },
})

// helper function for table sorting
const getSorting = (order, orderBy) =>
  order === "desc"
    ? (a, b) => (b[orderBy] < a[orderBy] ? -1 : 1)
    : (a, b) => (a[orderBy] < b[orderBy] ? -1 : 1)

type Props = {
  /** Material-UI styling */
  classes: Object,
  /** GOA data fetched from QuickGO API */
  goaData: Array<Object>,
}

type State = {
  /** The order to sort the column */
  order: string,
  /** The item to be ordered by */
  orderBy: string,
}

/**
 * The display table used inside each panel in the GO tabs.
 */

class DisplayTable extends Component<Props, State> {
  state = {
    order: "asc",
    orderBy: "evidence_code",
  }

  handleRequestSort = (event, property) => {
    const orderBy = property
    let order = "desc"

    if (this.state.orderBy === property && this.state.order === "desc") {
      order = "asc"
    }

    this.setState({ order, orderBy })
  }

  render() {
    const { order, orderBy } = this.state
    const { classes, goaData } = this.props

    // get array for item.attributes, helps for sorting table
    const goaDataAttributes = goaData.map((item: Object) => item.attributes)

    return (
      <Paper className={classes.root}>
        <Table>
          <colgroup>
            <col style={{ width: "25%" }} />
            <col style={{ width: "10%" }} />
            <col style={{ width: "20%" }} />
            <col style={{ width: "25%" }} />
            <col style={{ width: "15%" }} />
            <col style={{ width: "10%" }} />
          </colgroup>
          <EnhancedTableHead
            order={order}
            orderBy={orderBy}
            onRequestSort={this.handleRequestSort}
          />
          <TableBody>
            {goaDataAttributes
              .sort(getSorting(order, orderBy))
              .map((item: Object, index: number) => (
                <TableRow className={classes.row} key={index}>
                  <TableCell component="th" scope="row">
                    {qualifierFormatter(item.qualifier)} {item.goterm}
                  </TableCell>
                  <TableCell>
                    <a
                      className={classes.link}
                      href={evidenceLinkGenerator(item.evidence_code)}
                      target="_blank">
                      {item.evidence_code}
                    </a>
                  </TableCell>
                  <TableCell>
                    {item.with !== null &&
                      item.with.map((item: Object) =>
                        item.connectedXrefs.map((xref: Object, i: string) => (
                          <Fragment key={i}>
                            <span>
                              <a
                                className={classes.link}
                                href={withLinkGenerator(xref.id)}
                                target="_blank">
                                {xref.db}:{xref.id}
                              </a>
                            </span>
                            <br />
                          </Fragment>
                        )),
                      )}
                  </TableCell>
                  <TableCell>
                    {item.extensions !== null &&
                      item.extensions.map((item: Object) =>
                        item.connectedXrefs.map((xref: Object, i: string) => (
                          <Fragment key={i}>
                            <span>
                              {" "}
                              <em>{removeUnderscores(xref.relation)}</em>{" "}
                              <a
                                className={classes.link}
                                href={withLinkGenerator(xref.id)}
                                target="_blank">
                                ({xref.db}:{xref.id})
                              </a>
                            </span>
                            <br />
                          </Fragment>
                        )),
                      )}
                  </TableCell>
                  <TableCell>
                    <a
                      className={classes.link}
                      href={pubLinkGenerator(item.publication)}
                      target="_blank">
                      {item.publication}
                    </a>
                  </TableCell>
                  <TableCell>{dateConverter(item.date)}</TableCell>
                </TableRow>
              ))}
          </TableBody>
        </Table>
      </Paper>
    )
  }
}

export default withStyles(styles)(DisplayTable)
