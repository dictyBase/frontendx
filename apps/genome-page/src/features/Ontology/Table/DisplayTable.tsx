import React from "react"
import { makeStyles } from "@material-ui/core/styles"
import Table from "@material-ui/core/Table"
import TableBody from "@material-ui/core/TableBody"
import TableCell from "@material-ui/core/TableCell"
import TableRow from "@material-ui/core/TableRow"
import Paper from "@material-ui/core/Paper"
// import EnhancedTableHead from "./EnhancedTableHead"
import ExtensionsDisplay from "./ExtensionsDisplay"
import WithDisplay from "./WithDisplay"
import pubLinkGenerator from "../utils/pubLinkGenerator"
import evidenceLinkGenerator from "../utils/evidenceLinkGenerator"
import dateConverter from "../utils/dateConverter"
import qualifierFormatter from "../utils/qualifierFormatter"
import sourceLinkGenerator from "../utils/sourceLinkGenerator"
// import getSorting from "./utils/getSorting"
import { GeneGOA } from "common/@types/gene-data"

const useStyles = makeStyles({
  root: {
    width: "100%",
    overflowX: "auto",
  },
  row: {
    "&:nth-of-type(even)": {
      // backgroundColor: theme.palette.background.default,
    },
  },
  link: {
    textDecoration: "none",
    color: "#4C5E81",
    "&:visited": {
      color: "#4C5E81",
    },
    "&:hover": {
      textDecoration: "underline",
    },
  },
  head: {
    backgroundColor: "#e6f2ff",
  },
  headerCell: {
    color: "#333",
    fontWeight: 600,
  },
})

type Props = {
  /** GO Annotations */
  data: Array<GeneGOA>
}

/**
 * The display table used inside each panel in the GO tabs.
 */

const DisplayTable = ({ data }: Props) => {
  const classes = useStyles()

  return (
    <Paper className={classes.root}>
      <Table>
        <colgroup>
          <col style={{ width: "25%" }} />
          <col style={{ width: "20%" }} />
          <col style={{ width: "10%" }} />
          <col style={{ width: "15%" }} />
          <col style={{ width: "10%" }} />
          <col style={{ width: "10%" }} />
          <col style={{ width: "10%" }} />
        </colgroup>
        {/* <EnhancedTableHead
            order={goa.tableOrder}
            orderBy={goa.tableSortBy}
            onRequestSort={this.handleRequestSort}
          /> */}
        <TableBody>
          {data
            // .sort(getSorting(goa.tableOrder, goa.tableSortBy))
            .map((item: GeneGOA, index: number) => (
              <TableRow className={classes.row} key={index}>
                <TableCell component="th" scope="row">
                  {qualifierFormatter(item.qualifier)} {item.go_term}
                </TableCell>

                <TableCell>
                  <ExtensionsDisplay extensions={item.extensions} />
                </TableCell>

                <TableCell>
                  <a
                    className={classes.link}
                    href={evidenceLinkGenerator(item.evidence_code)}
                    target="_blank"
                    rel="noopener noreferrer">
                    {item.evidence_code}
                  </a>
                </TableCell>

                <TableCell>
                  <WithDisplay withData={item.with} />
                </TableCell>

                <TableCell>
                  <a
                    className={classes.link}
                    href={pubLinkGenerator(item.publication)}
                    target="_blank"
                    rel="noopener noreferrer">
                    {item.publication}
                  </a>
                </TableCell>

                <TableCell>{dateConverter(item.date)}</TableCell>

                <TableCell>
                  <a
                    className={classes.link}
                    href={sourceLinkGenerator(item.assigned_by)}
                    target="_blank"
                    rel="noopener noreferrer">
                    {item.assigned_by}
                  </a>
                </TableCell>
              </TableRow>
            ))}
        </TableBody>
      </Table>
    </Paper>
  )
}

export default DisplayTable
