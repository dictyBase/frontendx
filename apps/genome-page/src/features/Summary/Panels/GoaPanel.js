// @flow
import React from "react"
import Skeleton from "react-loading-skeleton"
import { withStyles } from "@material-ui/core/styles"
import Table from "@material-ui/core/Table"
import TableBody from "@material-ui/core/TableBody"
import TableCell from "@material-ui/core/TableCell"
import TableRow from "@material-ui/core/TableRow"
import Paper from "@material-ui/core/Paper"

const styles = theme => ({
  root: {
    width: "100%",
  },
  table: {
    borderBottom: "1px dotted #A3BAE9",
  },
  tableLeftData: {
    backgroundColor: "#DFE8F6",
    minWidth: 150,
    maxWidth: 200,
  },
  tableRightData: {
    width: "80%",
  },
})

type Props = {
  /** Material-UI styling */
  classes: Object,
  /** Object representing the "goa" slice of state */
  panelData: Object,
}

/**
 * Panel to display Gene Ontology Annotations in Gene Summary tab
 */

const GoaPanel = (props: Props) => {
  const { classes, panelData } = props

  if (panelData.error) {
    return (
      <div>
        <br />
        <p>Sorry! There was an error loading the items.</p>
        <br />
      </div>
    )
  }

  if (panelData.isFetching) {
    return (
      <div>
        <br />
        <Skeleton count={10} />
      </div>
    )
  }

  // set variables that represent filtered arrays for use in each row
  const molecular = panelData.data.data.filter(
    item => item.type === "molecular_function",
  )
  const biological = panelData.data.data.filter(
    item => item.type === "biological_process",
  )
  const cellular = panelData.data.data.filter(
    item => item.type === "cellular_component",
  )

  return (
    <Paper className={classes.root}>
      <Table className={classes.table}>
        <TableBody>
          <TableRow>
            <TableCell
              component="th"
              scope="row"
              className={classes.tableLeftData}>
              Molecular Function
            </TableCell>
            <TableCell className={classes.tableRightData}>
              {molecular.map((item: Object, i: string) => {
                if (molecular.length === i + 1) {
                  return `${item.attributes.goterm} (${
                    item.attributes.evidence_code
                  })`
                }
                return `${item.attributes.goterm} (${
                  item.attributes.evidence_code
                }), `
              })}
            </TableCell>
          </TableRow>
          <TableRow>
            <TableCell
              component="th"
              scope="row"
              className={classes.tableLeftData}>
              Biological Process
            </TableCell>
            <TableCell className={classes.tableRightData}>
              {biological.map((item, i) => {
                if (biological.length === i + 1) {
                  return `${item.attributes.goterm} (${
                    item.attributes.evidence_code
                  })`
                }
                return `${item.attributes.goterm} (${
                  item.attributes.evidence_code
                }), `
              })}
            </TableCell>
          </TableRow>
          <TableRow>
            <TableCell
              component="th"
              scope="row"
              className={classes.tableLeftData}>
              Cellular Component
            </TableCell>
            <TableCell className={classes.tableRightData}>
              {cellular.map((item, i) => {
                if (cellular.length === i + 1) {
                  return `${item.attributes.goterm} (${
                    item.attributes.evidence_code
                  })`
                }
                return `${item.attributes.goterm} (${
                  item.attributes.evidence_code
                }), `
              })}
            </TableCell>
          </TableRow>
        </TableBody>
      </Table>
    </Paper>
  )
}

export default withStyles(styles)(GoaPanel)
