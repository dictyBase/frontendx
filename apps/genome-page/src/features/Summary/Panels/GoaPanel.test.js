import React from "react"
import { mount } from "enzyme"
import "../../../setupTests"
import GoaPanel from "./GoaPanel"
import Table from "@material-ui/core/Table"
import TableRow from "@material-ui/core/TableRow"
import TableCell from "@material-ui/core/TableCell"

describe("Summary/Panels/GoaPanel", () => {
  const props = {
    classes: {},
    panelData: {
      data: {
        data: [
          [
            {
              type: "biological_process",
              id: "GO:0000281",
              attributes: {
                date: "20120522",
                evidence_code: "IMP",
                goterm: "mitotic cytokinesis",
                qualifier: "acts_upstream_of_or_within",
                publication: "PMID:12499361",
                with: null,
                extensions: null,
                assigned_by: "dictyBase",
              },
            },
            {
              type: "biological_process",
              id: "GO:0000902",
              attributes: {
                date: "20051219",
                evidence_code: "IMP",
                goterm: "cell morphogenesis",
                qualifier: "acts_upstream_of_or_within",
                publication: "PMID:12499361",
                with: null,
                extensions: null,
                assigned_by: "dictyBase",
              },
            },
            {
              type: "molecular_function",
              id: "GO:0005515",
              attributes: {
                date: "20110913",
                evidence_code: "IPI",
                goterm: "protein binding",
                qualifier: "enables",
                publication: "PMID:21441344",
                with: [{ connectedXrefs: [{ db: "UniProtKB", id: "Q54HG2" }] }],
                extensions: null,
                assigned_by: "dictyBase",
              },
            },
            {
              type: "cellular_component",
              id: "GO:0005938",
              attributes: {
                date: "20120522",
                evidence_code: "IDA",
                goterm: "cell cortex",
                qualifier: "part_of",
                publication: "PMID:12499361",
                with: null,
                extensions: null,
                assigned_by: "dictyBase",
              },
            },
            {
              type: "cellular_component",
              id: "GO:0015629",
              attributes: {
                date: "20120522",
                evidence_code: "IDA",
                goterm: "actin cytoskeleton",
                qualifier: "part_of",
                publication: "PMID:21441344",
                with: null,
                extensions: null,
                assigned_by: "dictyBase",
              },
            },
          ],
        ],
      },
    },
  }
  const wrapper = mount(<GoaPanel {...props} />)

  describe("initial render", () => {
    it("renders without crashing", () => {
      wrapper
    })

    it("always renders one Table", () => {
      expect(wrapper.find(Table).length).toBe(1)
    })

    it("always renders three TableRows", () => {
      expect(wrapper.find(TableRow).length).toBe(3)
    })

    it("always renders six TableCells", () => {
      expect(wrapper.find(TableCell).length).toBe(6)
    })

    it("receives two props", () => {
      expect(Object.keys(wrapper.props()).length).toBe(2)
    })
  })
})
