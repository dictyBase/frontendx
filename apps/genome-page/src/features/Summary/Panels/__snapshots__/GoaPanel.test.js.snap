// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matching a snapshot of Summary/Panels/GoaPanel 1`] = `
<div
  className="MuiPaper-root-5 MuiPaper-elevation2-9 MuiPaper-rounded-6 GoaPanel-root-1"
>
  <table
    className="MuiTable-root-32 GoaPanel-table-2"
  >
    <tbody
      className="MuiTableBody-root-33"
    >
      <tr
        className="MuiTableRow-root-34"
      >
        <th
          aria-sort={null}
          className="MuiTableCell-root-39 MuiTableCell-body-41 GoaPanel-tableLeftData-3"
          scope="row"
        >
          Molecular Function
        </th>
        <td
          aria-sort={null}
          className="MuiTableCell-root-39 MuiTableCell-body-41 GoaPanel-tableRightData-4"
          scope={undefined}
        >
          protein binding (IPI)
        </td>
      </tr>
      <tr
        className="MuiTableRow-root-34"
      >
        <th
          aria-sort={null}
          className="MuiTableCell-root-39 MuiTableCell-body-41 GoaPanel-tableLeftData-3"
          scope="row"
        >
          Biological Process
        </th>
        <td
          aria-sort={null}
          className="MuiTableCell-root-39 MuiTableCell-body-41 GoaPanel-tableRightData-4"
          scope={undefined}
        >
          mitotic cytokinesis (IMP), 
          cell morphogenesis (IMP)
        </td>
      </tr>
      <tr
        className="MuiTableRow-root-34"
      >
        <th
          aria-sort={null}
          className="MuiTableCell-root-39 MuiTableCell-body-41 GoaPanel-tableLeftData-3"
          scope="row"
        >
          Cellular Component
        </th>
        <td
          aria-sort={null}
          className="MuiTableCell-root-39 MuiTableCell-body-41 GoaPanel-tableRightData-4"
          scope={undefined}
        >
          cell cortex (IDA), 
          actin cytoskeleton (IDA)
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
